### DISCUSSION TITLE
Structure guide / best practices

### DISCUSSION THREAD

**Discussion by iron3oxide** [lowdefy/lowdefy#1236] [Ideas and feature requests]
While I enjoy the flexibility of lowdefy, I think it would be great to have some kind of guideline on how to best structure a non-trivial lowdefy project. Experienced teams will inevitably come up with some form of guideline for themselves, but someone just starting out might be tempted to put everything in two very messy yaml files and call it a day.

There are of course boundaries to how opinionated and low-level such a guideline should be (even though I for one would welcome it if someone told me that I have to put `layout` before `properties` or vice versa); some liberty is always beneficial. 

I tried to formulate such a guide myself by looking at the example apps, but frankly, they don't seem to follow a very consistent guideline either (which is totally understandable and to be fair, the case management system seems well structured). 

Some questions I asked myself over the past few days, where I wished for such a guide:

- I put modals in separate files and needed their block_id to open them from another file, which meant I had to _ref them in that file first. Where do I put that _ref? It's not arbitrary since a closed modal inside a blocks array will change the layout a bit. Should I avoid putting modals in separate files for that very reason?
- Where do I best define something like `onMount` actions? E.g. in the button, the box that encloses it, the card that encloses the box or in the page? If it depends: What are the determining factors?
- When does it make sense to put a component into its own file? Only if I want to reuse it? Or also if my file gets too long/messy? Is it purely subjective when and if a file is too long/messy or are there empirically/anecdotally suggested thresholds, like the ~90 characters a line of code should max. entail?

Maybe some of these questions have easy or obvious answers that I simply didn't see yet, but I still think some form of guide would be very helpful, not just for beginners. Don't you have one at lowdefy?

**Comment by Gervwyk** [comment]
I 100% agree with you it has been a topic of discussion in our team often, we've scaled out a few projects and tried different approaches, I think we are getting closer to a "recommended" approach. @SamTolmay has some notes on this. We need to standardize our projects and will do so while transitioning them to v4, by which time we'll publish the notes.

But for now, let mw add this; the _ref operator can be confusing since it can be used for two purposes like you mentioned:
1. to template config for reusability,
2. to break up config for readability / navigation.

When building a project and creating files, intentionally decide for which reason you are doing it. Most often it is 2, when it is 1, be careful, most likely the simplicity of copy and paste beats the complexity of abstraction.

We usually break pages into folders with requests and components sub folders for readability. With a shared folder for templates.

**Reply by iron3oxide** [reply]
That's a good start, I instinctively used a components sub folder for my project as well and grouped the contents into sub folders by types ("boxes", "cards" etc.). I haven't come to the requests part of my project yet, so "templates" is placed along those sub folders, but your suggestions makes sense and will probably have saved me the minute between realizing I need to template requests as well and restructuring my project accordingly. Now I'm even more excited for v4!
